sql/partition_info.cc

```diff
@@ -1209,8 +1209,8 @@ void partition_info::vers_update_col_vals(THD *thd, partition_element *el0, part
       if (*val_item < t)
       {
         val_item->set_time(&t);
-        col_val->fixed= 0;
       }
+      col_val->fixed= 0;
```

Comment:
Can you describe what exactly this fixes? Why you decided to refix `col_val` when it was not updated? Looks like ad hoc solution. Where `Field_timestampf` was constructed, where it was destroyed? I don't see such info in analysis.

Reply:
Here is some info. Hope it's suitable to get the picture.

> Why you decided to refix col_val when it was not updated? Looks like ad hoc solution.

Yes, you're right. It just fixes one of several observed errors: reading freed memory.

> Where Field_timestampf was constructed, where it was destroyed?

We don't care because we don't touch it's memory. We touch a memory which this instance points to: `TABLE::record[0]`

Btw, 3 kinds of timestamp comparison problems happens also in `partition.test`. But tests pass :(

Do you have some thoughts on how to do a proper fix?

Reply:
> Btw, 3 kinds of timestamp comparison problems happens also in partition.test. But tests pass :(

That's partially incorrect: no freed memory read in `partition.test`.

Reply:
This is not `TABLE::record[0]`: `Field_timestampf::ptr` can point to any arbitrary buffer. This is general scheme to fix premature allocation:

1. Find where the buffer was allocated;
2. Find where the buffer was freed;
3. Fix allocation: use longer living pool.

Reply:
https://github.com/tempesta-tech/mariadb/issues/441#issuecomment-355291226

`thd->mem_root` is ok for `column_value` of other types of partitioning but not enough for versioning partition.

Here is patch with another `mem_root` which is in use in `partition_info.cc`. Assertion failure disappears. What do you think about it?
```diff
diff --git a/sql/partition_info.cc b/sql/partition_info.cc
index e1e83db7b31..d102d5bf379 100644
--- a/sql/partition_info.cc
+++ b/sql/partition_info.cc
@@ -2972,7 +2972,8 @@ bool partition_info::fix_column_value_functions(THD *thd,
         }
         thd->got_warning= save_got_warning;
         thd->variables.sql_mode= save_sql_mode;
-        if (!(val_ptr= (uchar*) thd->memdup(field->ptr, len)))
+        if (!(val_ptr= (uchar *)(vers_info ? memdup_root(&table->s->mem_root, field->ptr, len)
+                                           : thd->memdup(field->ptr, len))))
         {
           mem_alloc_error(len);
           result= TRUE;
```

Reply:
This certainly should not be `TABLE_SHARE::mem_root` as it is inter-thread.
Try this:
```diff
--- a/sql/partition_info.h
+++ b/sql/partition_info.h
@@ -562,7 +562,10 @@ class partition_info : public Sql_alloc
       Field *f= part_field_array[i];
       bitmap_set_bit(f->table->write_set, f->field_index);
     }
+    MEM_ROOT *old_root= thd->mem_root;
+    thd->mem_root= &table->mem_root;
     result= check_range_constants(thd, false);
+    thd->mem_root= old_root;
     vers_info->stat_serial= table->s->stat_serial;
     mysql_rwlock_unlock(&table->s->LOCK_stat_serial);
     return result;
```
or find connection-wide **mem_root**.

Reply:
That worked. PR updated.